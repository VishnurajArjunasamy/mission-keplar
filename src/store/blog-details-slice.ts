import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { BlogIF, BlogWithIdIF } from "../modals/blog-list-modal";
import { toast } from "react-toastify";
import { set } from "firebase/database";
import { updateBlogs } from "../services/blog-add-update";
import { ValidationErrors } from "../modals/new-blog-modal";

export const updateBlog = createAsyncThunk(
  "updateBlog",
  async (blog: BlogWithIdIF[]) => {
    return updateBlogs(blog);
  }
);

interface BlogDetailsState {
  isReadOnly: boolean;
  error: string | null | ValidationErrors;
  loading: boolean;
}

const initialState: BlogDetailsState = {
  error: null,
  loading: false,
  isReadOnly: true,
};

const blogDetailsSlice = createSlice({
  name: "blogDetails",
  initialState,
  reducers: {
    setIsReadOnly(state, action) {
      state.isReadOnly = action.payload;
    },
    setError(state, action) {
      state.error = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(updateBlog.pending, (state) => {
        state.loading = true;
      })
      .addCase(updateBlog.fulfilled, (state, action) => {
        state.loading = false;
        toast.success("Blog updated successfully");
      })
      .addCase(updateBlog.rejected, (state, action) => {
        state.loading = false;
        state.error = (action.payload as string) || "Error while updating blog";
        toast.error(state.error);
      });
  },
});

export const { setIsReadOnly, setError: setUpdateBlogError } =
  blogDetailsSlice.actions;
export default blogDetailsSlice.reducer;
