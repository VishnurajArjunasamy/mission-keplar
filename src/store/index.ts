import { configureStore } from "@reduxjs/toolkit";
import sideBarReducer from "./side-bar-slice";
import blogListReducer from "./blog-list-slice";
import newBlogReducer from "./new-blog-slice";

const store = configureStore({
  reducer: {
    sideBar: sideBarReducer,
    blogList: blogListReducer,
    newBlog: newBlogReducer,
  },
});
export default store;

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
