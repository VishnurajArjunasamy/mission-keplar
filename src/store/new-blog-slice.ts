import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { BlogIF, BlogWithIdIF } from "../modals/blog-list-modal";
import { postNewBlog } from "../services/new-blog";
import { ValidationErrors } from "../modals/new-blog-modal";
import { toast } from "react-toastify";


export const addNewBlog = createAsyncThunk(
  "newBlog/addNewBlog",
  async (newBlog: BlogWithIdIF[]) => {
    return postNewBlog(newBlog);
  }
);

interface NewBlogState {
  loading: boolean;
  error: string | null | ValidationErrors;
  isAdded: boolean;
  showNewBlogModal: boolean;
}

const initialState: NewBlogState = {
  loading: false,
  error: null,
  isAdded: false,
  showNewBlogModal: false,
};

const newBlogSlice = createSlice({
  name: "newBlog",
  initialState,
  reducers: {
    setShowNewBlogModal: (state, action) => {
      state.showNewBlogModal = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(addNewBlog.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(addNewBlog.fulfilled, (state, action) => {
        state.loading = false;
        toast.success("New blog added successfully");
      })
      .addCase(addNewBlog.rejected, (state, action) => {
        state.loading = false;
        state.error =
          (action.payload as string) || "Error while adding new blog";
        toast.error(state.error);
      });
  },
});

export const { setShowNewBlogModal, setError: setNewBlogError } =
  newBlogSlice.actions;
export default newBlogSlice.reducer;
