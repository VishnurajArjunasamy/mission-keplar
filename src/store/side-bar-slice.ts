import { createSlice } from "@reduxjs/toolkit";
import { SideBarState } from "../modals/sideBarModal";

const inititalState: SideBarState = {
  showMembers: false,
  isDarkMode: false,
};

const sideBarSlice = createSlice({
  name: "SideBar",
  initialState: inititalState,
  reducers: {
    toggleShowMembers: () => {},
    toggleDarkMode: () => {},
  },
});

export const { toggleDarkMode, toggleShowMembers } = sideBarSlice.actions;

export default sideBarSlice.reducer;
