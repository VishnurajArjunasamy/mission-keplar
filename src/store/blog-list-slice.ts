import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { BlogListState } from "../modals/blog-list-modal";
import { fetchBlogList } from "../services/blog-list";
import { getAvailableFilters } from "../utils/getAvailableFilters";
import { ERROR_MSG } from "../constants/app.constants";
import { addNewBlog } from "./new-blog-slice";
import { updateBlog } from "./blog-details-slice";

export const fetchBlogs = createAsyncThunk("blogList/fetchData", async () => {
  return fetchBlogList();
});

export const initialState: BlogListState = {
  data: null,
  loading: false,
  error: null,
  bufferBlogId: null,
  selectedBlog: null,
  searchQuery: null,
  filters: {},
};

const blogListSlice = createSlice({
  name: "blogList",
  initialState,
  reducers: {
    changeFilter: (state, action) => {
      const id: string = action.payload;
      state.filters[id] = !state.filters[id];
    },
    setSelectedBlogs: (state, action) => {
      state.selectedBlog = action.payload;
    },
    setBufferBlogId: (state, action) => {
      state.bufferBlogId = action.payload;
    },
    setSearchQuery: (state, action) => {
      state.searchQuery = action.payload;
    },
    setBlogs: (state, action) => {
      state.data = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchBlogs.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchBlogs.fulfilled, (state, action) => {
        state.loading = false;
        state.data = action.payload;
        state.selectedBlog = action.payload[0].id;
        state.filters = getAvailableFilters(action.payload);
      })
      .addCase(fetchBlogs.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || ERROR_MSG.MESSAGE;
      })
      .addCase(addNewBlog.fulfilled, (state, action) => {
        state.data = action.payload;
      })
      .addCase(updateBlog.fulfilled, (state, action) => {
        state.data = action.payload;
      });
  },
});

export const {
  changeFilter,
  setSelectedBlogs,
  setSearchQuery,
  setBlogs,
  setBufferBlogId,
} = blogListSlice.actions;
export default blogListSlice.reducer;
