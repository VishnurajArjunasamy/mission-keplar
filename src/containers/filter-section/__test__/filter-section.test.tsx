import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import FilterSection from "../filter-section";
import { changeFilter } from "../../../store/blog-list-slice";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

const mockStore = configureStore();

describe("FilterSection Component", () => {
  let store: ReturnType<typeof mockStore>;

  beforeEach(() => {
    store = mockStore({
      blogList: {
        filters: {
          Technology: true,
          Lifestyle: false,
          Health: true,
        },
      },
      sideBar: {
        isDarkMode: false,
      },
    });
    store.dispatch = jest.fn();
  });

  it("renders filter checkboxes with correct initial state", () => {
    render(
      <Provider store={store}>
        <FilterSection />
      </Provider>
    );

    expect(screen.getByLabelText("Technology")).toBeInTheDocument();
    expect(screen.getByLabelText("Lifestyle")).toBeInTheDocument();
    expect(screen.getByLabelText("Health")).toBeInTheDocument();

    expect(screen.getByLabelText("Technology")).toBeChecked();
    expect(screen.getByLabelText("Lifestyle")).not.toBeChecked();
    expect(screen.getByLabelText("Health")).toBeChecked();
  });

  it("dispatches changeFilter action when a checkbox is clicked", () => {
    render(
      <Provider store={store}>
        <FilterSection />
      </Provider>
    );

    const lifestyleCheckbox = screen.getByLabelText("Lifestyle");

    fireEvent.click(lifestyleCheckbox);

    expect(store.dispatch).toHaveBeenCalledWith(changeFilter("Lifestyle"));
  });

  it("applies the correct style based on isDarkMode", () => {
    store = mockStore({
      blogList: {
        filters: {
          Technology: true,
          Lifestyle: false,
          Health: true,
        },
      },
      sideBar: {
        isDarkMode: true,
      },
    });

    render(
      <Provider store={store}>
        <FilterSection />
      </Provider>
    );
    const filterSection = screen.getByText("FILTER").closest("div");
    expect(filterSection).toHaveClass("dark");
  });

  it("Matches Snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <FilterSection />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
