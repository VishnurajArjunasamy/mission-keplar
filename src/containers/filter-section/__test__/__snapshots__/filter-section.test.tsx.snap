// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSection Component Matches Snapshot 1`] = `
<DocumentFragment>
  <div
    class="filterSection light"
  >
    <h1>
      FILTER
    </h1>
    <div
      class="checkBoxWrapper light"
    >
      <input
        checked=""
        id="Technology"
        type="checkbox"
      />
      <label
        for="Technology"
      >
        Technology
      </label>
    </div>
    <div
      class="checkBoxWrapper light"
    >
      <input
        id="Lifestyle"
        type="checkbox"
      />
      <label
        for="Lifestyle"
      >
        Lifestyle
      </label>
    </div>
    <div
      class="checkBoxWrapper light"
    >
      <input
        checked=""
        id="Health"
        type="checkbox"
      />
      <label
        for="Health"
      >
        Health
      </label>
    </div>
  </div>
</DocumentFragment>
`;
